server/src/main/java/org/opensearch/tasks/CancellableTask.java

```diff
@@ -71,10 +71,14 @@ public CancellableTask(
     /**
      * This method is called by the task manager when this task is cancelled.
      */
-    public void cancel(String reason) {
+    public void cancel(String message) {
+        assert message != null;
+        cancel(new Reason(message));
+    }
+
+    public void cancel(Reason reason) {
         assert reason != null;
-        if (cancelled.compareAndSet(false, true)) {
-            this.reason = reason;
+        if (reasonReference.compareAndSet(null, reason)) {
```

Comment:
CAS uses object references for comparison (`==`), not values (`.equals` method).

The statement `reasonReference.compareAndSet(null, reason)` means that atomic set operation will only succeed if the previous value was `null`. If this statement is executed again (either with logically or referentially equivalent Reason, doesn't matter), the previous value would no longer be `null` and the atomic set would fail.

As you rightly mentioned, this is to ensure that Reason can only be set once. `SetOnce` can also be used here, in fact, it is a wrapper on top of `AtomicReference` itself which uses the same `compareAndSet(null, object)` instruction.

Reply:
Previously this set-once behaviour was achieved with an AtomicBoolean (`cancelled`) surrounding a volatile string (`reason`). This change replaces them with a single `AtomicReference` to a `Reason` object.